mlir-sycl/lib/Dialect/SYCL/Analysis/MemoryAccessAnalysis.cpp

```diff
@@ -0,0 +1,652 @@
+  ++nRows;
+  data.resize(nRows * nColumns);
+  return nRows - 1;
+}
+
+size_t MemoryAccessMatrix::appendRow(ArrayRef<Value> elems) {
+  size_t row = appendRow();
+  setRow(row, elems);
+  return row;
+}
+
+SmallVector<Value> MemoryAccessMatrix::getColumn(size_t column) const {
+  assert(column < nColumns && "the matrix must contain the given column");
+  SmallVector<Value> columnCopy;
+  columnCopy.reserve(nRows);
+  for (size_t row = 0; row < nRows; ++row)
+    columnCopy.emplace_back(at(row, column));
+  return columnCopy;
+}
+
+void MemoryAccessMatrix::setColumn(size_t col, ArrayRef<Value> elems) {
+  assert(col < nColumns && "the matrix must contain the given column");
+  assert(elems.size() == nRows && "elems size must match column length!");
+  for (size_t row = 0; row < nRows; ++row)
+    at(row, col) = elems[row];
+}
+
+void MemoryAccessMatrix::fillColumn(size_t col, Value val) {
+  assert(col < nColumns && "the matrix must contain the given column");
+  for (size_t row = 0; row < nRows; ++row)
+    at(row, col) = val;
+}
+
+void MemoryAccessMatrix::fill(Value val) {
+  for (size_t row = 0; row < nRows; ++row)
+    for (size_t col = 0; col < nColumns; ++col)
+      at(row, col) = val;
+}
+
+MemoryAccessMatrix MemoryAccessMatrix::getRows(std::set<size_t> rows) const {
+  assert(llvm::all_of(rows, [this](size_t row) { return row < nRows; }) &&
+         "out of bounds");
+  MemoryAccessMatrix subMatrix(rows.size(), nColumns);
+  size_t pos = 0;
+  for (size_t rowNum : rows)
+    subMatrix.setRow(pos++, getRow(rowNum));
+  return subMatrix;
+}
+
+MemoryAccessMatrix
+MemoryAccessMatrix::getColumns(std::set<size_t> columns) const {
+  assert(llvm::all_of(columns,
+                      [this](size_t column) { return column < nColumns; }) &&
+         "out of bounds");
+  MemoryAccessMatrix subMatrix(nRows, columns.size());
+  size_t pos = 0;
+  for (size_t colNum : columns)
+    subMatrix.setColumn(pos++, getColumn(colNum));
+  return subMatrix;
+}
+
+MemoryAccessMatrix
+MemoryAccessMatrix::getSubMatrix(std::set<size_t> rows,
+                                 std::set<size_t> columns) const {
+  return getRows(std::move(rows)).getColumns(std::move(columns));
+}
+
+bool MemoryAccessMatrix::isSquare() const { return (nRows == nColumns); }
+
+bool MemoryAccessMatrix::isZero(DataFlowSolver &solver) const {
+  return (llvm::all_of(data,
+                       [&solver](Value val) { return ::isZero(val, solver); }));
+}
+
+bool MemoryAccessMatrix::isDiagonal(DataFlowSolver &solver) const {
+  if (!isSquare())
+    return false;
+
+  for (size_t row = 0; row < nRows; ++row)
+    for (size_t col = 0; col < nColumns; ++col) {
+      Value val = at(row, col);
+      bool isOnDiagonal = (row == col);
+
+      // All values on the diagonal must be non-zero.
+      if (isOnDiagonal && ::isZero(val, solver))
+        return false;
+      // All other values must be zero.
+      if (!isOnDiagonal && !::isZero(val, solver))
+        return false;
+    }
```

Comment:
```suggestion
  for (size_t row = 0; row < nRows; ++row) {
    for (size_t col = 0; col < nColumns; ++col) {
      Value val = at(row, col);
      bool isOnDiagonal = (row == col);

      // All values on the diagonal must be non-zero.
      if (isOnDiagonal && ::isZero(val, solver))
        return false;
      // All other values must be zero.
      if (!isOnDiagonal && !::isZero(val, solver))
        return false;
    }
  }
```
This helps readability